base/errorshow.jl

```diff
@@ -208,6 +208,32 @@ function show_convert_error(io::IO, ex::MethodError, @nospecialize(arg_types_par
     end
 end
 
+"""
+    methoderror_hints
+
+Packages can provide hints about appropriate corrective actions for
+specific `MethodError`s with `push!(Base.methoderror_hints, hintmessage)`, where
+`hintmessage` should be a function like
```

Comment:
Does it makes sense to expose this API as a function, like `pushdisplay` and `atreplinit`? It can be used as something like

```julia
register_methoderror_hint() do f, arg_types, kwargs
    ...
end
```

or just `register_methoderror_hint(hintmessage)`. It seems to be helpful to disentangle the API from the internal representation so that future improvements are easier.

Reply:
I thought about that. I decided against it (but am happy to reconsider) because it would be a fair number of exports If all our exception types end up admitting hints.

Reply:
Hmm, we could do it with just one: `register_error_hint(handler, MethodError)`. We'd have to document the expected arguments of `handler` for each error type separately. Thoughts?

Reply:
That makes sense.  Then how about

```julia
register_error_hint(MethodError) do f, arg_types, kwargs
    ...
end
```

?

Reply:
Ah, race condition.

Reply:
For example, the inability to customize `InexactError` messages is one of the problems in FixedPointNumbers.jl. However, I think it is a problem of `InexactError` and should not be generalized here. Under the principle that type piracy should be prohibited, I think only `MethodError` is the special case.

Reply:
I don't think there is any type piracy going on here. I think a similar hook system can be provided for other exception types if necessary, without any type piracy.

(Type piracy occurs in the method dispatch because there is only one method you can dispatch to. Hook systems like this allow multiple matches. That's why I think type piracy is irrelevant here.)

Reply:
> I don't think there is any type piracy going on here. 

I agree with you, because the type piracy should be prohibited.
In the principle, module `Foo` have the right to control the errors which depend only on the `FooType` defined in the the module `Foo`. Thus, if you don't like the default error message, `Foo` can define a new `Exception` type as needed to show its own error message. However, `MethodError` cannot be controlled by `Foo` for now, even if it depends only on `FooType`.

Reply:
Yes, for error _messages_, I agree that's the right way to do it. This API is about _hints_. It is for the situation multiple third parties can provide useful information to the user. Though I cannot find other exception types that are useful to have such entry-point ATM.